src/RestApi/StoreApi/Routes/Cart.php

```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Cart route.
+ *
+ * @package WooCommerce/Blocks
+ */
+
+namespace Automattic\WooCommerce\Blocks\RestApi\StoreApi\Routes;
+
+defined( 'ABSPATH' ) || exit;
+
+use Automattic\WooCommerce\Blocks\RestApi\StoreApi\Utilities\CartController;
+
+/**
+ * Cart class.
+ */
+class Cart extends AbstractRoute {
+	/**
+	 * Get the namespace for this route.
+	 *
+	 * @return string
+	 */
+	public function get_namespace() {
+		return 'wc/store';
+	}
+
+	/**
+	 * Get the path of this REST route.
+	 *
+	 * @return string
+	 */
+	public function get_path() {
+		return '/cart';
+	}
+
+	/**
+	 * Get method arguments for this REST route.
+	 *
+	 * @return array An array of endpoints.
+	 */
+	public function get_args() {
+		return [
+			[
+				'methods'  => \WP_REST_Server::READABLE,
+				'callback' => [ $this, 'get_response' ],
+				'args'     => [
+					'context' => $this->get_context_param( [ 'default' => 'view' ] ),
+				],
+			],
+			'schema' => [ $this->schema, 'get_public_item_schema' ],
+		];
+	}
+
+	/**
+	 * Handle the request and return a valid response for this endpoint.
+	 *
+	 * @param \WP_REST_Request $request Request object.
+	 * @return \WP_Error|\WP_REST_Response
+	 */
+	public function get_response( \WP_REST_Request $request ) {
+		$controller = new CartController();
+		$cart       = $controller->get_cart_instance();
+
+		if ( ! $cart || ! $cart instanceof \WC_Cart ) {
+			return new \WP_Error( 'woocommerce_rest_cart_error', __( 'Unable to retrieve cart.', 'woo-gutenberg-products-block' ), array( 'status' => 500 ) );
+		}
```

Comment:
I realize we throw an error here because `\WP_Rest_Server` handles WP_Error, not exceptions.  If we have a common `get_response` callback wrapper on the `AbstractRoute`, it could have a single try catch that converts caught exceptions to `WP_Error`.  Then all our logic can just throw exceptions.  That way in this particular case (and anywhere the `CartController` is invoked, you could have the exception happen in the `CartController::get_cart_instance` method and that would save duplicating this same error logic everywhere (i.e. centralize the exception where it occurs, if `get_cart_instance` can't retrieve a cart instance, it throws).

So `AbstractRoute`, would have something like:

```php
public function get_response( \WP_REST_Request $request ) {
    $response = null;
    try { 
        $response = $this->get_route_response( $request );
    } catch ( WC_Rest_Exception $error ) {
        $response = new WP_Error(
          $error->code,
          $error->message,
          [ 'status' => $error->status ]
        )
    } catch ( Exception $error ) {
      $response = new WP_Error(
        'unknown_server_error',
        $error->message,
        [ 'status' => $error->status ]
      )
    }
    return $response;
}

abstract protected function get_route_response( $request );
```

Reply:
@nerrad So because we have a callback per method, the only way to do something like this would be to have a try catch in each and a handler in abstract.

However, by using the request method we can have a single callback like you suggested. I ran with that. Just updated - can you take a look?